### PR TITLE
[WIP] - SCM Webhook

### PR DESCRIPTION
**Pull request by chimpuz**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### CODE REVIEW

File: jazz_scm-webhook/index.js [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+				rejectUnauthorized: false
+			};
+			request(options, function (error, response, body) {
+				if (error) {
+					logger.error('Error invoking service: ' + error);
+					return cb(JSON.stringify(errorHandler.throwInternalServerError("Unexpected Error occured,"+ error)));
+				} else {
+					if(response.statusCode === 200){
+						var output = {
+							message: 'successfully recorded bitbucket activity to jazz_events.',
```
Comment (suryajak) [inline jazz_scm-webhook/index.js]: Update message to say git

File: jazz_scm-webhook/index.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+							event_id : body.data.event_id
+						};
+						logger.info("OutPut: " +JSON.stringify(output));
```
Comment (suryajak) [inline jazz_scm-webhook/index.js]: combine logger.info

File: jazz_scm-webhook/swagger/swagger.json [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+            "COMMIT_CODE",
+            "DELETE_BRANCH",
+            "RAISE_PR",
+            "MERGE_PR",
+            "DECLINE_PR",
+            "UPDATE_PR",
+            "COMMENT_PR"
+          ],
+          "type": "string",
+          "description": "Describes the action that is taking place in the Bitbucket."
```
Comment (suryajak) [inline jazz_scm-webhook/swagger/swagger.json]: Update description to say SCM

File: jazz_scm-webhook/index.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+			result.branch = changes.fromRef.branch.name;
+			result.prlink = body.pullrequest.link;
+			result.target = changes.toRef.branch.name;
+		} else if(body.push){
+			changes = body.push.changes[0];	
+		}
+		
+		switch (value) {
+			case 'repo:push': 					
+				if(changes.created && !changes.closed && changes.old === null && changes.new !== null) {
```
Comment (suryajak) [inline jazz_scm-webhook/index.js]: Check if changes is not null. Also combine and use short form:
`if(changes && changes.created && !changes.closed && !changes.old && changes.new) {`

File: jazz_scm-webhook/index.js [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+					result.branch = changes.new.name;
+					if(type === 'tag'){
+						result.event_name = 'CREATE_TAG';
+					} else if(type === 'branch' && changes.new.name === 'master'){
+						result.event_name = 'COMMIT_TEMPLATE';
+						result.event_type = config.event_type.onboarding;
+					} else {
+						result.event_name = 'CREATE_BRANCH';
+					}					
+				} else if(!changes.created && changes.closed){
```
Comment (suryajak) [inline jazz_scm-webhook/index.js]: check if changes is not null

File: jazz_scm-webhook/index.js [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+					var objtype = changes.old.type;
+					result.branch = changes.old.name;
+					if(objtype === 'tag'){
+						result.event_name = 'DELETE_TAG';
+					} else {
+						result.event_type = config.event_type.deletion;
+						result.event_name = 'DELETE_BRANCH';
+					} 		
+				} else if(!changes.created && !changes.closed && changes.old !== null){
```
Comment (suryajak) [inline jazz_scm-webhook/index.js]: check if changes it no null

File: jazz_scm-webhook/config/stg-config.json [hunk 7]
```diff
@@ -0,0 +1,64 @@
+{
+  "events_url": "https://{conf-apikey}.execute-api.{conf-region}.amazonaws.com/stg/jazz/events",
+  "event_status": {
+    "completed": "COMPLETED"
+  },
+  "event_handler": {
+    "bitbucket": "BITBUCKET"
```
Comment (suryajak) [inline jazz_scm-webhook/config/stg-config.json]: Key name should be SCM? Also this should be replaced as the service is getting built.

File: jazz_scm-webhook/index.js [hunk 8]
```diff
@@ -0,0 +1,180 @@
+/**
+API to record SCM activity
+@author: Aanand12
```
[the remark this answers is not in the document]
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Done

File: jazz_scm-webhook/Jenkinsfile [hunk 9]
```diff
@@ -0,0 +1,52 @@
+#!groovy
+import groovy.json.JsonOutput
+import groovy.transform.Field
+
+// To be replaced as @Field def repo_credential_id = "value" except for config_loader
+@Field def repo_credential_id = "jenkins1cred"
```
[the remark this answers is not in the document]
Comment (chimpuz) [reply jazz_scm-webhook/Jenkinsfile]: Done

File: jazz_scm-webhook/Jenkinsfile [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  stage('Triggering the Jenkins build') {
+    withCredentials([[$class: 'UsernamePasswordMultiBinding', credentialsId: config_loader.JENKINS.CREDENTIAL_ID, passwordVariable: 'PWD', usernameVariable: 'UNAME']]){
+    echo "Build triggered via branch::${env.BRANCH_NAME}"
+
+    def build_job = config_loader.API.API_BUILD_URI
+    def job_url = JenkinsLocationConfiguration.get().getUrl() + build_job
+
+    sh "curl -X GET -k -v -u \"$UNAME:$PWD\"  \"" + job_url + "&service_name=scm-webhook&domain=test&scm_branch=${env.BRANCH_NAME}\""
```
[the remark this answers is not in the document]
Comment (chimpuz) [reply jazz_scm-webhook/Jenkinsfile]: Done

File: jazz_scm-webhook/index.js [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+module.exports.handler = (event, context, cb) => {
+
+	//Initializations
+	var errorHandler = errorHandlerModule();
+	var config = configObj(event);
+	logger.init(event, context);
+	logger.info("Webhook-events:" + JSON.stringify(event));
+	
+	if(event === undefined || event.body === undefined ){
```
[the remark this answers is not in the document]
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Done

File: jazz_scm-webhook/index.js [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+		logger.error("Events is empty! so unable to find the scm activity!");
+		return cb(JSON.stringify(errorHandler.throwInputValidationError("Unable to find the bitbucket activity!")));
```
[the remark this answers is not in the document]
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Done

File: jazz_scm-webhook/index.js [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+					'pr_link' : servContext.prlink,
+					'target' : servContext.target,
+					'hash' : servContext.hash
+				}
+			};
+		var possibleEventName = config.event_name,
+			isvalidEventName = false;
+		for (var idx in possibleEventName ){
+			if ((servContext.event_name).toLowerCase() == possibleEventName[idx].toLowerCase()){
+				isvalidEventName = true;
```
[the remark this answers is not in the document]
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Done

File: jazz_scm-webhook/index.js [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+			} 
+		}
+		
+		logger.info("bodyObj::" +JSON.stringify(bodyObj));
```
[the remark this answers is not in the document]
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Done

File: jazz_scm-webhook/index.js [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+		logger.info("config.events_url::" + config.events_url);
+
+		if(isvalidEventName){
+			//Invoking platform events api to record bit bucket activity.
```
[the remark this answers is not in the document]
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Done

File: jazz_scm-webhook/index.js [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+					'cache-control': 'no-cache',
+					'content-type': 'application/json' 
+				},
+				body : bodyObj,
+				json: true,
+				rejectUnauthorized: false
+			};
+			request(options, function (error, response, body) {
+				if (error) {
+					logger.error('Error invoking service: ' + error);
```
[the remark this answers is not in the document]
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Done

File: jazz_scm-webhook/index.js [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+					return cb(JSON.stringify(errorHandler.throwInternalServerError("Unexpected Error occured,"+ error)));
```
[the remark this answers is not in the document]
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Done

File: jazz_scm-webhook/index.js (hunk 1 above)
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Done

File: jazz_scm-webhook/index.js (hunk 2 above)
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Removed

File: jazz_scm-webhook/swagger/swagger.json (hunk 3 above)
Comment (chimpuz) [reply jazz_scm-webhook/swagger/swagger.json]: Done

File: jazz_scm-webhook/index.js (hunk 4 above)
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Done

File: jazz_scm-webhook/index.js (hunk 5 above)
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Done

File: jazz_scm-webhook/index.js (hunk 6 above)
Comment (chimpuz) [reply jazz_scm-webhook/index.js]: Done

File: jazz_scm-webhook/config/stg-config.json (hunk 7 above)
Comment (chimpuz) [reply jazz_scm-webhook/config/stg-config.json]: This will change in next iteration when there is GitLabs added.

**Comment by chimpuz** [comment]
Incoprorated PR comment

**Pull request merged**